File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        // ret maxDepth, subAnswer
        
        Pair<Integer, Integer> fa = helper(root);
        return Math.max(fa.getKey() - 1, fa.getValue());
    }
    public Pair<Integer, Integer> helper(TreeNode root) {
        if (root == null) {
            return new Pair(0,0);
        }
        Pair<Integer, Integer> la = helper(root.left);
        Pair<Integer, Integer> ra = helper(root.right);
        int md = (Math.max(la.getKey(), ra.getKey()) + 1);
        int sa = (Math.max(la.getValue(), ra.getValue()));
        int sah = la.getKey() + ra.getKey();
        // if (la.getKey() != 0) {
        //     sah += la.getKey();
        // }
        // if (ra.getKey() != 0) {
        //     sah += ra.getKey();
        // }
        sa = Math.max(sa, sah);
        return new Pair(md, sa);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        return height(root) != -1
    }
    fun height(root: TreeNode?): Int {
        if (root == null) { 
            return 0
        }
        val hl = height(root.left)
        val hr = height(root.right)
        if (hl == -1 || hr == -1 || abs(hl - hr) > 1) {
            return -1
        }
        return maxOf(hl, hr) + 1
    }
}
